File: packages/client/src/types/events.ts
/**
 * Streaming event types for SSE and WebSocket communications
 */

import type { UsageStats } from "./common";

/**
 * Execution events streamed during task execution
 */
export type ExecutionEvent =
  | { type: "text"; content: string }
  | { type: "thinking"; content: string }
  | { type: "tool_call_start"; id: string; name: string; input: object }
  | { type: "tool_call_delta"; id: string; input_delta: object }
  | { type: "tool_call_end"; id: string }
  | {
      type: "tool_result";
      id: string;
      success: boolean;
      output?: object;
      error?: string;
    }
  | {
      type: "artifact";
      id: string;
      name: string;
      mime_type: string;
      size_bytes: number;
    }
  | {
      type: "progress";
      percentage: number;
      step: string;
      step_number: number;
      total_steps: number;
    }
  | { type: "error"; code: string; message: string; recoverable: boolean }
  | { type: "done"; usage: UsageStats };

/**
 * Tool stream events for streaming tool invocations
 */
export type ToolStreamEvent =
  | { type: "output"; data: object }
  | { type: "progress"; percentage: number; message: string }
  | { type: "error"; code: string; message: string }
  | { type: "done"; success: boolean; duration_ms: number };

/**
 * Plan events for task planning updates
 */
export type PlanEvent =
  | { type: "task.added"; task: PlanTask }
  | { type: "task.updated"; task: PlanTask }
  | { type: "task.removed"; task_id: string }
  | { type: "plan.reordered"; task_ids: string[] };

export interface PlanTask {
  id: string;
  content: string;
  status: "pending" | "in_progress" | "completed";
  order: number;
}

/**
 * Conformance test events
 */
export type ConformanceEvent =
  | { type: "test.started"; test_id: string; test_name: string }
  | { type: "test.passed"; test_id: string }
  | { type: "test.failed"; test_id: string; error: string }
  | { type: "test.skipped"; test_id: string; reason: string }
  | { type: "progress"; completed: number; total: number }
  | { type: "done"; result: ConformanceResult };

export interface ConformanceResult {
  passed: number;
  failed: number;
  skipped: number;
  duration_ms: number;
}

/**
 * Harness-level events for event streaming
 */
export type HarnessEvent =
  | { type: "hook.triggered"; hook_id: string; event: string; data: object }
  | { type: "execution.started"; execution_id: string }
  | { type: "execution.completed"; execution_id: string; status: string }
  | { type: "session.created"; session_id: string }
  | { type: "session.ended"; session_id: string }
  | { type: "skill.installed"; skill_id: string }
  | { type: "skill.uninstalled"; skill_id: string }
  | { type: "error"; code: string; message: string };

/**
 * WebSocket client messages for interactive sessions
 */
export type SessionClientMessage =
  | { type: "message"; id: string; content: string }
  | { type: "stdin"; id: string; data: string }
  | { type: "cancel"; execution_id: string };

/**
 * WebSocket server messages for interactive sessions
 */
export type SessionServerMessage =
  | { type: "text"; id: string; content: string }
  | { type: "thinking"; id: string; content: string }
  | { type: "tool_call"; id: string; tool: string; input: object }
  | { type: "stdout"; id: string; data: string }
  | { type: "stderr"; id: string; data: string }
  | { type: "prompt"; id: string; prompt: string }
  | { type: "artifact"; id: string; name: string; mime_type: string }
  | { type: "error"; id: string; code: string; message: string }
  | { type: "done"; id: string; usage: UsageStats };

File: packages/python/src/openharness/types/events.py
"""
Event types for streaming execution.
"""

from typing import Any, Literal, Union

from pydantic import BaseModel, Field

from .common import UsageStats


class TextEvent(BaseModel):
    """Text content chunk."""
    type: Literal["text"] = "text"
    content: str


class ThinkingEvent(BaseModel):
    """Extended thinking content."""
    type: Literal["thinking"] = "thinking"
    thinking: str


class ToolCallStartEvent(BaseModel):
    """Tool invocation starting."""
    type: Literal["tool_call_start"] = "tool_call_start"
    id: str
    name: str
    input: dict[str, Any] = Field(default_factory=dict)


class ToolCallDeltaEvent(BaseModel):
    """Partial tool input update."""
    type: Literal["tool_call_delta"] = "tool_call_delta"
    id: str
    input_delta: dict[str, Any]


class ToolCallEndEvent(BaseModel):
    """Tool invocation complete."""
    type: Literal["tool_call_end"] = "tool_call_end"
    id: str


class ToolResultEvent(BaseModel):
    """Tool execution result."""
    type: Literal["tool_result"] = "tool_result"
    id: str
    success: bool
    output: Any | None = None
    error: str | None = None


class ProgressEvent(BaseModel):
    """Execution progress update."""
    type: Literal["progress"] = "progress"
    percentage: float = Field(ge=0, le=100)
    step: str | None = None
    step_number: int | None = None
    total_steps: int | None = None


class ErrorEvent(BaseModel):
    """Error during execution."""
    type: Literal["error"] = "error"
    code: str
    message: str
    recoverable: bool = False


class DoneEvent(BaseModel):
    """Execution complete."""
    type: Literal["done"] = "done"
    usage: UsageStats | None = None


class ArtifactEvent(BaseModel):
    """Artifact generated."""
    type: Literal["artifact"] = "artifact"
    id: str
    name: str
    content_type: str
    content: str | bytes


# Union of all execution events
ExecutionEvent = Union[
    TextEvent,
    ThinkingEvent,
    ToolCallStartEvent,
    ToolCallDeltaEvent,
    ToolCallEndEvent,
    ToolResultEvent,
    ProgressEvent,
    ErrorEvent,
    DoneEvent,
    ArtifactEvent,
]


# Tool-specific stream events
class ToolStreamStartEvent(BaseModel):
    """Tool stream starting."""
    type: Literal["tool_stream_start"] = "tool_stream_start"
    tool_id: str
    tool_name: str


class ToolStreamDataEvent(BaseModel):
    """Tool stream data chunk."""
    type: Literal["tool_stream_data"] = "tool_stream_data"
    tool_id: str
    data: Any


class ToolStreamEndEvent(BaseModel):
    """Tool stream complete."""
    type: Literal["tool_stream_end"] = "tool_stream_end"
    tool_id: str
    success: bool
    error: str | None = None


ToolStreamEvent = Union[
    ToolStreamStartEvent,
    ToolStreamDataEvent,
    ToolStreamEndEvent,
]
